### PR TITLE
1894 Fix Dashboard Date Filter

### PR DESCRIPTION
**Pull request by DaveBlueFrontier**

### CODE REVIEW

File: app/views/dashboards/index.html.haml [hunk 1]
```diff
@@ -80,6 +89,22 @@
           #chartAverageTestsTechnician{:style => "height: 200px; width: 100%;"}
             = react_component 'CanvasJsChart', chartDiv: 'chartAverageTestsTechnician', chartData: Extras::Charts::VerticalColumns.new(@dashboard_report.average_tests_per_technician).render
 
+    -#%section.row
+      .small-8.columns.box
+        %header
+          =image_tag('img/site.png')
+          %h3 Devices Not Responding
```
Comment (jorgegorka) [inline app/views/dashboards/index.html.haml]: String should be localised like the others.

File: app/views/dashboards/index.html.haml [hunk 2]
```diff
@@ -80,6 +89,22 @@
...
+        .box-content
+          #chartTotalTestsSite{:style => "height: 200px; width: 100%;"}
```
Comment (jorgegorka) [inline app/views/dashboards/index.html.haml]: There is another chart with this Id.

File: app/views/dashboards/index.html.haml [hunk 3]
```diff
@@ -80,6 +89,22 @@
...
+            = react_component 'CanvasJsChart', chartDiv: 'chartDevicesNotResponding', chartData: Extras::Charts::HorizontalColumns.new(@dashboard_report.devices_not_reporting).render
```
Comment (jorgegorka) [inline app/views/dashboards/index.html.haml]: Method is called not_reporting and title of the chart says not responding

File: app/views/dashboards/index.html.haml [hunk 4]
```diff
@@ -80,6 +89,22 @@
...
+      .small-4.columns.box
+        %header
+          = image_tag('img/user.png')
+          %h3 Device Error Codes
```
Comment (jorgegorka) [inline app/views/dashboards/index.html.haml]: String should be localised like the others.

File: app/views/dashboards/index.html.haml [hunk 5]
```diff
@@ -80,6 +89,22 @@
...
+        .box-content
+          #chartAverageTestsTechnician{:style => "height: 200px; width: 100%;"}
```
Comment (jorgegorka) [inline app/views/dashboards/index.html.haml]: There is another chart with this Id.

File: lib/reports/outstanding_orders.rb [hunk 6]
```diff
@@ -20,7 +20,8 @@ def latest_encounter
 
       # TODO: , order results by time desc to remove a time check below:
       # filter['order_by']='-test.reported_time'
-      to_diff_time = filter['until'] || Time.now
+      to_diff_time = Time.now
+      to_diff_time = Time.parse(filter['until'].to_s) unless filter['until'].nil?
```
Comment (jorgegorka) [inline lib/reports/outstanding_orders.rb]: Why assigning a value to same variable two times?  I guess line 24 can be removed.

Comment (DaveBlueFrontier) [inline lib/reports/outstanding_orders.rb]: It has an unless option on it - so the line 23 sets the default value, line 24 will change it unless the filer is nil.

Comment (jorgegorka) [inline lib/reports/outstanding_orders.rb]: I didn't realised it.  You are right!!

Comment (DaveBlueFrontier) [inline lib/reports/outstanding_orders.rb]: No problem :)

File: app/views/dashboards/index.html.haml (hunk 1 above)
Comment (DaveBlueFrontier) [inline app/views/dashboards/index.html.haml]: Ok

File: app/views/dashboards/index.html.haml (hunk 2 above)
Comment (DaveBlueFrontier) [inline app/views/dashboards/index.html.haml]: Good spot.

This whole section is currently not active until the reports Extras has the data returning for it.

File: app/views/dashboards/index.html.haml (hunk 5 above)
Comment (DaveBlueFrontier) [inline app/views/dashboards/index.html.haml]: Good spot.

This whole section is currently not active until the reports Extras has the data returning for it.

File: app/views/dashboards/index.html.haml (hunk 3 above)
Comment (DaveBlueFrontier) [inline app/views/dashboards/index.html.haml]: I've updated the name.  It was based on the old name the system used.

Methods: devices_not_responding  and  error_codes_by_device  currently have no data returns in lib/reports/dashboard.rb.

I was leaving this until the appropriate methods were in place in Reports::

File: lib/reports/outstanding_orders.rb (hunk 6 above)
Comment (jorgegorka) [inline lib/reports/outstanding_orders.rb]: Happy to merge once tests are green.  Thanks!!!

File: lib/reports/dashboard.rb [hunk 7]
```diff
@@ -32,4 +32,13 @@ def total_tests_by_device
   def average_tests_per_technician
     Reports::AverageTechnicianTests.new(@current_user, @navigation_context, @options).generate_chart
   end
+
+  def devices_not_reporting
+    nil
```
Comment (tripledub) [inline lib/reports/dashboard.rb]: No need to return `nil` def devices_not_reporting; end` will do for a placeholder method

File: lib/reports/dashboard.rb [hunk 8]
```diff
@@ -32,4 +32,13 @@ def total_tests_by_device
...
+  end
+
+  def error_codes_by_device
+    nil
```
Comment (tripledub) [inline lib/reports/dashboard.rb]: see above

File: lib/reports/dashboard.rb (hunk 7 above)
Comment (tripledub) [inline lib/reports/dashboard.rb]: actually are we not just querying mysql for this? In which case this doesn't really belong here

Comment (DaveBlueFrontier) [inline lib/reports/dashboard.rb]: I thought that it was done like this so as to provide a consistent flow for the system.  Dashboard index calls lib/reports/dashboard.rb, which in turn calls on the Reports:: so as to get the data in a standardised format.  Makes the whole system more modular and less tied to methods that are in different locations.  Removes some of the spaghettification maybe!

Comment (jorgegorka) [inline lib/reports/dashboard.rb]: Those methods are fine. reports::dashboard is an object that generates all the reports needed on the dashboard (you can check all the code needed in the index method before and after this object was created.).  Every individual report is generated in its own class and then reports::dashboard calls them.  I think they are ok.

**Pull request merged**